Respect the force flag on refspecs in git_remote_fetch

alexjg opened the pull request:
This fixes #2957 

If the `->force` flag on a refspec is not set then we check that the new refs are descendants of the current refs before updating them.

ethomson commented:
@carlosmn any chance you can spare some 👀 for this?  Your knowledge of the networking code is far superior to mine.

carlosmn commented:
Oops I guess we've been behaving like everything is forced forever. I let the tests run and it looks like there's memory errors here where one of the variables isn't initialized. Running it locally it keeps failing at 

    remote::fetch::do_update_refs_if_not_descendant_and_force [/home/carlos/git/libgit2/tests/remote/fetch.c:75]

While the overall approach seems like it should work, there do seem to be some issues in the code that can be caught by the tests.

carlosmn commented:
I suspect the error happens in the case where we first create the remote-tracking branch so we look it up but `old` is left unwritten-to. Making the graph check conditional on `!error` should solve this.

alexjg commented:
I notice that the test I wrote is also failing 

```
 1:   1) Failure:
1: remote::fetch::do_update_refs_if_not_descendant_and_force [../tests/remote/fetch.c:76]
1:   Function call failed: (git_commit_create(commit1id, repo1, REPO1_REFNAME, sig, sig, ((void *)0), "one", empty_tree, 0, ((void *)0)))
1:   error -15 - failed to create commit: current tip is not the first parent
```

This is strange as it passes on my machine. It looks like it's failing to clean up after the first test and so the test repo is in an unexpected state. Am I missing something on the cleanup?

carlosmn commented:
It is most likely something to do with cleanup or the interaction between tests. Running just that failing test does pass.

carlosmn commented:
When you create the new root commit the second time around, the branch has commits, so the empty parent list means a zero-commit as first parent, whereas the existing tip of the branch is whatever actual commit you created.

carlosmn commented:
This might be the result of #5842 which is somewhat recent and might explain why you don't see it yourself if you branched off of something earlier.

alexjg commented:
Ah interseting, I've rebased off `main` and now I get a test failure. Will see what I can do.

alexjg commented:
I had to expose a `cl_fs_rm` function to directly remove the test repo in the cleanup code. The tests pass now but this may not be the idiomatic way to remote the repos?

alexjg commented:
Hey peeps, any chance someone could re-run the tests for this?

carlosmn commented:
Exposing internal clar functionality is not the way to go. Check out how other tests, e.g. the `mkdir` tests, deal with custom cleanup https://github.com/libgit2/libgit2/blob/868f4bcb4d3290f4b5320f030fccdf1e7fc8ac8a/tests/core/mkdir.c#L6-L22

alexjg commented:
Thanks for the hint! I've reowrked this to use `git_futils_rmdir_r`

alexjg commented:
Hey folks, sorry to bother, any chance someone can run the tests again?

tiennou commented:
Sure, CI has been kicked into action.

alexjg commented:
I managed to reproduce the memory sanitizer failure in a local version of the CI pipeline and I've fixed the failure there. This should be good to go now.

alexjg commented:
Could someone nudge CI into action again?

alexjg commented:
Looks like all the tests are passing, is there anything else to do to merge this?

ethomson commented:
@carlosmn is this looking good?  :shipit:

carlosmn commented on tests/remote/fetch.c:
```diff
@@ -0,0 +1,184 @@
+	git_buf repo2_path_buf = GIT_BUF_INIT;
+	const char *sandbox = clar_sandbox_path();
+
+	cl_git_pass(git_buf_join(&repo1_path_buf, '/', sandbox, "fetchtest_repo1"));
```
`git_buf_joinpath` would be nicer here; it's what I was expecting and made me think you were trying to prepend slash :/

alexjg replied:
Done

carlosmn reviewed (approved):
The code itself seems fine, mostly just the tests should be closer to the patterns and utilities that we already use elsewhere.

The test functions themselves could be made smaller and the "time-traveling fetch" function could also perform the check but that's minor.

carlosmn commented:
One bigger change and consequence of this is that the `update_tips` callback should also be called when it doesn't get updated so it's able to report that to the user, as I believe git also does.

alexjg commented on tests/remote/fetch.c:
```diff
@@ -0,0 +1,184 @@
+		strncpy(result, path, strlen(path));
+		return result;
+	}
+}
```
:facepalm: removed

alexjg commented on tests/remote/fetch.c:
```diff
@@ -0,0 +1,184 @@
+
+	cl_git_pass(git_buf_join(&repo1_path_buf, '/', sandbox, "fetchtest_repo1"));
+	repo1_path = calloc(repo1_path_buf.size, sizeof(char));
+	git_buf_copy_cstr(repo1_path, repo1_path_buf.size, &repo1_path_buf);
```
Done

alexjg commented:
@carlosmn  Thanks for the review, I believe I've addressed your comments. Regarding the larger change to `update_tips`, that would imply an extension of the callback interface which could be done in a subsequent PR right?

carlosmn commented:
> Regarding the larger change to `update_tips`, that would imply an extension of the callback interface which could be done in a subsequent PR right?

Yes, that would require increasing the version of the remote callback structure and adding some sort of `update_tips_ext` to include whether the update succeeded or not and why.

ethomson commented:
:shipit:  Thanks for the change - and the great test @alexjg - and for the code review @carlosmn.
